pkg/agent/agent.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/prometheus/prometheus/pkg/exemplar"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+
+	"github.com/prometheus/prometheus/tsdb/record"
+	"github.com/prometheus/prometheus/tsdb/wal"
+)
+
+// Options to create WAL.
+type WALOptions struct {
+	BaseDir     string
+	SegmentSize units.Base2Bytes
+	Compression bool
+}
+
+// Implements the Storage interface for agent to write metrics to WAL.
+type WALDelegate struct {
+	logger log.Logger
+	wal    *wal.WAL
+}
+
+//
+type WALAppender struct {
+	logger   log.Logger
+	delegate *WALDelegate
+	series   []record.RefSeries
+	samples  []record.RefSample
+}
+
+var (
+	ErrUnsupported = errors.New("unsupported operation with WAL-only storage")
+)
+
+// Create a new WAL delegate for writing metrics to WAL.
+func NewWALDelegate(logger log.Logger, wOptions WALOptions) (storage.Storage, error) {
+	walDir := wOptions.BaseDir + string(os.PathSeparator) + "wal"
+	walLog, err := wal.NewNotify(logger, prometheus.DefaultRegisterer, walDir, true, int(wOptions.SegmentSize), wOptions.Compression)
+	if err != nil {
+		return nil, err
+	}
+	walDelegate := &WALDelegate{
+		logger: logger,
+		wal:    walLog,
+	}
+	go truncateWALSegment(logger, walLog)
+	return walDelegate, nil
+}
+
+// Handles WAL notifications for checkpoint and truncate WAL segements.
+func truncateWALSegment(logger log.Logger, walLog *wal.WAL) {
+	// TODO: better condition
+	for true {
+		select {
+		case walSegmentNum := <-wal.CWalNotify:
+			if walSegmentNum >= 0 {
+				keep := func(id uint64) bool {
+					return false
+				}
+				// Checkpoint a WAL segment.
+				_, err := wal.Checkpoint(logger, walLog, walSegmentNum, walSegmentNum, keep, 0)
+				if err != nil {
+					level.Error(logger).Log("msg", "Checkpointing a segment is failed. ",
+						" : Segment #: ", walSegmentNum, " error: ", err.Error())
+				} else {
+					// Truncate a WAL segment.
+					err = walLog.Truncate(walSegmentNum)
+					if err != nil {
+						level.Error(logger).Log("msg", "Truncating a segment is failed. ",
+							" : Segment #: ", walSegmentNum, " error: ", err.Error())
+					}
+				}
+			}
+		default:
+			//TODO.
+		}
+	}
+}
+
+func (w *WALDelegate) Querier(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
+	return nil, nil
+}
+
+func (w *WALDelegate) ChunkQuerier(ctx context.Context, mint, maxt int64) (storage.ChunkQuerier, error) {
+	return nil, nil
+}
+
+func (w *WALDelegate) Appender(ctx context.Context) storage.Appender {
+	return &WALAppender{
+		delegate: w,
+		logger:   w.logger,
+	}
+}
+
+func (w *WALDelegate) StartTime() (int64, error) {
+	startTime := time.Now().Unix() * 1000
+	return startTime, nil
+}
+
+func (w *WALDelegate) Close() error {
+	//TODO
+	return nil
+}
+
+// Appends series, samples.
+func (a *WALAppender) Append(ref uint64, l labels.Labels, t int64, v float64) (uint64, error) {
+	a.series = append(a.series, record.RefSeries{
+		Ref:    uint64(l.Hash()),
```

Comment:
Ah, this is a neat idea. Unfortunately, it is going to cause issues if there's a hash collision between two different series, which is definitely possible at scale. 

Also, I'm concerned about the performance of this if we'll be writing a series record for every sample. You might want to consider writing bechnmarks to do some comparisons on both the read (remote_write) and write (scrape) path.

Reply:
Looking into it more, I'm more convinced this won't work now once there's a hash collision. The QueueManager receives all [series records](https://github.com/prometheus/prometheus/blob/039b6514506260bd57619b9b26265aff771e4a61/storage/remote/queue_manager.go#L636) first, followed by all [sample records](https://github.com/prometheus/prometheus/blob/039b6514506260bd57619b9b26265aff771e4a61/storage/remote/queue_manager.go#L499).

This means if two series with the same hash get written within the WAL read period (10ms), only the label set of the newer series will be used for writing both samples.

Reply:
@rfratto this was a great catch.   Fixed this issue with a simple counter for series which gets reset for every new Segment.  

> How does this relate/differ from @rfratto implementation?

1. Lifecycle  of WAL remains the same (implementation) for both Prometheus and Prometheus-agent. Storage delegate functions to WAL are part of the new package `pkg/agent`
2. In the Prometheus agent mode, WAL emits Lifecycle events like Segment creation, remoteWriting is done processing a Segment. These events are handled by the package `pkg/agent`.

Reply:
> Ah, this is a neat idea. Unfortunately, I think this may cause issues in remote_write if there's a hash collision between two different series, which is definitely possible at scale.
> 
> Also, I'm concerned about the performance of this if we'll be writing a series record for every sample. You might want to consider writing bechnmarks to do some comparisons on both the read (remote_write) and write (scrape) path.

Yes @rfratto benchmark tests (and of course functional tests) are important. To increase write perf., I am contemplating between (a) buffering based on # of time series (b)  to cap under memory size, indicated at cmd. line. I am leaning more towards (b) as agent would be running on every node, clear cap on resources making sense to me. WDYT ?

Reply:
Honestly I think this still problematic and at first glance I don't understand the reasoning or how this is going to work.

Writing a series record for every incoming sample is unnecessary and negates the reason for having series records in the first place. A series record exists once in the WAL, assuming the series is still active. Remote write caches series labels from a series record, and so each subsequent sample record that is read just needs to lookup the reference # in the cache.

In order for what you're doing here to work, the agent is going to have to lock remote writes series cache every time a new segment is created so that it can be cleared, and then completely repopulate that cache from the new segments records. This is going to cause a lot of unnecessary CPU cycles to be spent on writing those records, reading them, and clearing + repopulating the cache.

Reply:
@cstyan reusing series record makes sense and locking remote writes series cache (for each new segment) does not help from resource cycles point of view.   With this observation, I would +1 to the series and storage [agent implementation](Initial draft of prometheus-agent #8785) that @rfratto proposed to build separate binaries for `Prometheus` and `Prometheus-agent` from a single codebase.